### ISSUE TITLE
Instance overlaps with all other instances

### ISSUE THREAD

**Issue by szumixie** [agda/agda#7486]
Bug discovered by @DrpontAgon. The following code typechecks in Agda 2.6.4.3 but instance resolution fails in 2.7.0:
```agda
data D : Set where
  con : D

record R : Set where
  no-eta-equality -- no error if eta-equality
  field proj : D

F : (D → Set) → D → Set -- no error if F : Set → D → Set
F k con = k con         -- no error if F k con = D

postulate
  instance i : {r : R} → F (λ _ → D) (R.proj r)

postulate
  A : Set
  instance a : A

it : {{A}} → A
it {{x}} = x

x : A
x = it
{-
Failed to solve the following constraints:
  Resolve instance argument _11 : A
  Candidates
    i : {r : R} → F (λ _ → D) (R.proj r)
    a : A
    (stuck)
-}
```
The instance `i` here is useless since instance arguments of that type can't be resolved, but it overlaps with all other instances and breaks instance resolution everywhere.

**Comment by szumixie** [comment]
Bisection points to aa5d8bbe723ea095d2af92ccaa2cfd2fbd89570a by @plt-amy.

**Comment by szumixie** [comment]
A more straightforward example that passes in 2.6.4.3 but fails in 2.7.0:
```agda
postulate
  A : Set

data D : Set where
  con : D

F : D → Set
F con = A

postulate
  instance a : A
  instance i : {d : D} → F d

it : {{A}} → A
it {{x}} = x

x : A
x = it
```
Removing `instance a : A` makes it fail in 2.6.4.3 and pass in 2.7.0. Maybe the new behavior is more correct.

**Comment by andreasabel** [comment]
Thanks for the report!

I tried both examples with Agda 2.5.1.1 - 2.6.4.3 (all 2.5 and 2.6 versions I have installed), and they pass.
So at least the release notes are wrong: https://github.com/agda/agda/blob/95eec65856ca2b21e1b6769b23b3c1cc7556d5ba/doc/release-notes/2.7.0.md?plain=1#L183-L184
So it looks like an unintended change that we should evaluate.

**Comment by plt-amy** [comment]
This change *was* unintended, but I think I'd file it more under "happy accident" than "mistake"... 

> Removing instance a : A makes it fail in 2.6.4.3 and pass in 2.7.0. Maybe the new behavior is more correct.

Agreed! These instances *do* overlap (in your second comment, we can pick either just fine). In the original comment, we're in a tricky situation: if `R` is no-eta, then we have a constraint

```agda
  F (λ _ → D) (R.proj _r) =< A
```

It'll be important here that `F` is injective, so we'll try to simplify the left-hand-side by matching the eliminations in `F _k con` (from the clause) against those in `F (λ _ → D) (_r .proj)` (from the equation). We're then in one of these situations:

- If everything is as in your original comment, then we can solve `_k x := D`, but we can't make any progress on `_r .proj = con`, so `F (λ _ → D) (_r .proj)` does not reduce, and injectivity gives up (though we do learn about the blocker `_r`). This means that the instance isn't discarded (yet), though since it is okay-for-overlap, it *could* be discarded by one of the pragmas...
- If we instead have `F k con = D`, then injectivity fails eagerly, since `D ≠ A` — i.e. none of the clauses have a head which matches `A` — so the instance is discarded.
- If `R` is instead `eta-equality`, then we can make progress on `_r .proj = con`, by solving `_r := record { proj = _α }` and `_α := con`. Injectivity succeeds, and we simplify the original equation to `D =< A`, which fails, so the instance is discarded.
- If `F` is instead `F : Type → D → Type`, `F k con = k`, then injectivity can once again fail eagerly, since (having instantiated the inverse with the eliminations `D`, `_r .proj`) we can see that no clause has a head matching `A`, so the instance is discarded.

So the situation in the OP is a very strange (and fragile!) one indeed! As far as the discrimination tree is concerned, `F (λ _ → D) (_r .proj)` *could* reduce to any other type at all, so we file it as an instance for "`_`", leaving it to the conversion checker to discard it at each use site. But when we get to a particular use site, the conversion checker kickflips onto a rake and gets stuck! The combination of the proper match stuck on a projection from a no-eta record, and having an `UnknownHead` in the inverse map, means that we well and truly can't make any progress here.

As for why it worked before?

```
io ~/d/P/agda % [...]/Agda-2.6.4/build/agda/agda Test159.agda -vtc.instance:60 -vtc.decl.instance:10
Checking Test159 (/home/amelia/default/Projects/agda/Test159.agda).
adding definition Test159.i to instance table for Test159.F
adding definition Test159.a to instance table for Test159.A
```

Old instance search was very happy to treat `i` as an instance for the "class" `F`.

**Comment by plt-amy** [comment]
> Old instance search was very happy to treat i as an instance for the "class" F.

And the reason for this, even though `F` is reducible (and so would never appear as the head of a type we're trying to *search for*), is that, when the old instance search was trying to figure out what `Def` to file an instance under, it would've seen `F (λ _ → D) (@0 .proj)`, which is `NotBlocked ReallyNotBlocked` (I think, since it's blocked on a variable — even though `@0` is out of scope — [no, seriously, we didn't go under the telescope before trying to reduce the head](https://github.com/agda/agda/blob/release-2.6.4-candidate/src/full/Agda/TypeChecking/Telescope.hs#L690-L691)), and decided that `F` is actually a normal form.

**EDIT**: Still going on about this, 2.6.4 rejects this program:

```agda
postulate
  A B : Set

F : Bool → Set
F true = A
F false = B

postulate
  instance i : {b : Bool} → F b

it : {A : Set} → {{A}} → A
it {{x}} = x

_ : A
_ = it

_ : B
_ = it
```

but the new, more accurate behaviour of potentially treating instance *heads* as reducible accepts it. It's basically *required* that we treat `i` as an instance "for any type" if we want this to succeed.

**Comment by andreasabel** [comment]
Agda dev meeting 2024-09-11: market this as a bug fix.
Add the example(s) to the testsuite.
